### ISSUE TITLE
Wrong Unicode codes for angle brackets

### ISSUE THREAD

**Issue by bytensky** [xkbcommon/libxkbcommon#47]
In file `src/keysym-utf.c`:
```
    { 0x0abc, 0x2329 }, /*            leftanglebracket 〈 LEFT-POINTING ANGLE BRACKET */
```
```
    { 0x0abe, 0x232a }, /*           rightanglebracket 〉 RIGHT-POINTING ANGLE BRACKET */
```
A the same time in file `keysymdef.h` of xproto:
```
#define XK_leftanglebracket              0x0abc  /*(U+27E8 MATHEMATICAL LEFT ANGLE BRACKET)*/
```
```
#define XK_rightanglebracket             0x0abe  /*(U+27E9 MATHEMATICAL RIGHT ANGLE BRACKET)*/
```
I think that U+27E8 and U+27E9 codes are more relevant for angle brackets.

Proposed diff file:
```
diff --git a/src/keysym-utf.c b/src/keysym-utf.c
index ffe2cea..5fdae24 100644
--- a/src/keysym-utf.c
+++ b/src/keysym-utf.c
@@ -534,9 +534,9 @@ static const struct codepair keysymtab[] = {
     { 0x0ab7, 0x215a }, /*                  fivesixths ⅚ VULGAR FRACTION FIVE SIXTHS */
     { 0x0ab8, 0x2105 }, /*                      careof ℅ CARE OF */
     { 0x0abb, 0x2012 }, /*                     figdash ‒ FIGURE DASH */
-    { 0x0abc, 0x2329 }, /*            leftanglebracket 〈 LEFT-POINTING ANGLE BRACKET */
+    { 0x0abc, 0x27e8 }, /*            leftanglebracket 〈 LEFT-POINTING ANGLE BRACKET */
     { 0x0abd, 0x002e }, /*                decimalpoint . FULL STOP */
-    { 0x0abe, 0x232a }, /*           rightanglebracket 〉 RIGHT-POINTING ANGLE BRACKET */
+    { 0x0abe, 0x27e9 }, /*           rightanglebracket 〉 RIGHT-POINTING ANGLE BRACKET */
     /*  0x0abf                                  marker ? ??? */
     { 0x0ac3, 0x215b }, /*                   oneeighth ⅛ VULGAR FRACTION ONE EIGHTH */
     { 0x0ac4, 0x215c }, /*                threeeighths ⅜ VULGAR FRACTION THREE EIGHTHS */
```

**Comment by bluetech** [comment]
Hmm... Looking at leftanglebracket

- The standard [here](https://www.x.org/releases/X11R7.7/doc/xproto/x11protocol.html#Legacy_KEYSYMs) does not specify any unicode value for it.

- The keysym list keysymdef.h in x11proto [here](https://cgit.freedesktop.org/xorg/proto/x11proto/tree/keysymdef.h) says U+27E9 MATHEMATICAL RIGHT ANGLE BRACKET in a comment.

- The keysym->unicode list in xkbcommon which comes from [here](https://www.cl.cam.ac.uk/~mgk25/ucs/keysym2ucs.c) has U+2329 LEFT-POINTING ANGLE BRACKET.

- The keysym->unicode list in Xlib [here](https://cgit.freedesktop.org/xorg/lib/libX11/tree/src/xlibi18n/imKStoUCS.c) has U+2039 SINGLE LEFT-POINTING ANGLE QUOTATION MARK.

A nice mess :)

I am not sure we should do here, but doing singular fixes ourselves sounds like a bad idea, it would only add yet another variant. That is, if we migrate one keysym we should migrate all of them.

@fooishbar Do you have any idea about this?

**Comment by fooishbar** [comment]
@bluetech Heh! No, no clue I'm afraid.

@bytensky How did you discover this issue?

**Comment by bytensky** [comment]
> @bytensky How did you discover this issue?

I was designing my own keyboard layout. I was looking for mnenomic names for U+27E8 and U+27E9 in [keysymdef.h](https://cgit.freedesktop.org/xorg/proto/x11proto/tree/keysymdef.h) and found names `leftanglebracket` and `rightanglebracket`. So I inserted these names in my layout file.

Than I started using my layout and found than symbols I typed was wrong.

**Comment by bluetech** [comment]
[According the Unicode] (found the reference from [here](https://en.wikipedia.org/wiki/Bracket_(mathematics)#Symbols_for_representing_angle_brackets)), the symbols we are currently using {LEFT,RIGHT}-POINTING ANGLE BRACKET are deprecated.

In light of this, I think your suggestion to switch to MATHEMATICAL LEFT ANGLE BRACKET makes sense. I'll apply your fix. Thanks!

**Comment by andrecbarros** [comment]
May I be wrong, but the unicode recommendations seems to be applied to math symbols. For punctuations the there exist both U+2039 (left single angle brackets) and U+203A (right single angle bracket) so, my suggestion is to add both symbols as XK_singleguillemotleft and XK_singleguillemotright. Note that this way we also cover these punctuation present on HTML5 charmap (and Windows-1252).

Regards,
André

**Comment by bluetech** [comment]
My best guesstimate is that the `{left,right}anglebracket` keysyms were meant for the mathematical symbols, so I think the change is good.

As for new keysyms, we don't really add them ourselves, they should go through xproto (now `xorg-proto`).

BTW these keysyms are somewhat deprecated anyway; nowadays there are `U1234`-style keysyms which can be used instead.

**Comment by andrecbarros** [comment]
Not really, they are quotation marks used on non-english languages (see http://jkorpela.fi/html/french.html). So, even though they look like the familiar math operators, that is not what they are.

Also, as I said, they are present in HTML5 charmap as quotation marks and may be in use by non-english languages as so.

My suggestion seems to apply.

**Comment by andrecbarros** [comment]
One more thing, I followed the used word on keysymdef.h that, actually, seems to be a glitch. The correct designation seems to be guillemet.

**Comment by bluetech** [comment]
Are you saying that `{left,right}anglebracket` should be defined as the single guillemet symbols? I don't think that's right, as I said, from my understanding, the original intention for the `{left,right}anglebracket` keysyms was to represent the mathematical symbols, hence we should choose the best codepoint for that.

It might make sense to add *new* keysyms for the single guillemet symbols, however this should be suggested for inclusion in xorg-proto first. But there is really no need to add new named keysyms these days, you should just use the unicode keysym instead. For example the latin `intl` variant does just that for single (half?) guillemet symbols (U2039/U203A):

https://cgit.freedesktop.org/xkeyboard-config/tree/symbols/latin?h=xkeyboard-config-2.23.1#n229

Does that make sense?

(BTW is it spelled "guillemet" like Wikipedia says, or "guillemot" like in the `guillemotleft` keysym? Confused!).

**Comment by andrecbarros** [comment]
No, {left,right}anglebacket should be left with the corrections you already did. I'm talking about missing symbols and misspelled designations.

**Comment by andrecbarros** [comment]
One more thing. It is not usually a good practice to left a glitch unnoticed so, if "synonyms" are not given (what would serve also as notices), at least a comment should be added to it.

**Comment by andrecbarros** [comment]
OK, reading my last post, it feels like I was lecturing about good practices. Sorry, it was not my intention.

Anyway, I'm aware that I could achieve the same results just using utf-8 strings but it would be nice to have the constants, if not more, just for reaching "completeness/parity" with HTML4 and Windows-1252 and, at same time, easing the processing of ordinary files from MS Windows machines.

May I help somehow, please, let me know.

Best regards

**Comment by bluetech** [comment]
> One more thing. It is not usually a good practice to left a glitch unnoticed so, if "synonyms" are not given (what would serve also as notices), at least a comment should be added about it.

Agreed! This reminds me of the HTTP [Referer header](https://en.wikipedia.org/wiki/HTTP_referer#Etymology) - a misspelling but set in stone.

Regarding new symbols and/or comments, you can look [here](https://github.com/xkbcommon/libxkbcommon/blob/master/scripts/makeheader) to see how we generate our header. The original source is [here](https://cgit.freedesktop.org/xorg/proto/xorgproto/tree/include/X11).

**Comment by andrecbarros** [comment]
Thanks for directions, will have a look at it.

Nice link to "referer", never payed attention to it before. :)

Best regards.
